lib/pyfrc/tests/docstring_test.py

```diff
@@ -1,140 +1,169 @@
+type_re = re.compile(r"^:type (\S+?):\s*(.+)$")
+both_re = re.compile(r"^:param (\S+?) (\S+?):\s*(.+)$")
 
 def ignore_object(o, robot_path):
     '''Returns true if the object can be ignored'''
-    
+
     if inspect.isbuiltin(o):
         return True
-    
+
     try:
         src = inspect.getsourcefile(o)
     except TypeError:
         return True
-    
+
     return src is None or not os.path.abspath(src).startswith(robot_path)
-        
+
 
 def print_fn_err(msg, parent, fn, errors):
-    
-    
     if inspect.isclass(parent):
-        name = '%s.%s' % (parent.__name__, fn.__name__) 
+        name = '%s.%s' % (parent.__name__, fn.__name__)
     else:
         name = '%s' % fn.__name__
-    
+    name = '%s.%s' % (parent.__name__, fn.__name__)
+
     err = "ERROR: %s '%s()'\n-> See %s:%s" % (msg, name,
                                               inspect.getsourcefile(fn),
                                               inspect.getsourcelines(fn)[1])
-    
-    print(err)
-    errors.append(err)
 
+    if err not in errors:
+        print(err, file=sys.stderr)
+        errors.append(err)
 
 
 def check_function(parent, fn, errors):
-    
     doc = inspect.getdoc(fn)
-    
+
     if doc is None:
         print_fn_err('No docstring for', parent, fn, errors)
-    
+
     elif pedantic_docstrings:
         # find the list of parameters
-        args, varargs, keywords, defaults = inspect.getargspec(fn)
+        args, varargs, keywords, defaults, _, _, annotations = inspect.getfullargspec(fn)
         if len(args) > 0 and args[0] == 'self':
             del args[0]
-            
+
         if varargs is not None:
             args.append(varargs)
-            
+
         if keywords is not None:
             args.append(keywords)
-            
+
         params = []
-               
-        for line in doc.splitlines():    
+
+        for line in doc.splitlines():
+            # :param arg: stuff
             match = param_re.match(line)
-            if not match:
-                continue
-            
-            arg = match.group(1)
-            if arg not in args:
-                print_fn_err("Param '%s' is documented but isn't a parameter for" % arg, parent, fn, errors)
-            
-            params.append(arg)
-            
+            if match:
+
+                arg = match.group(1)
+                if arg not in args:
+                    print_fn_err("Param '%s' is documented but isn't a parameter for" % arg, parent, fn, errors)
+
+                if arg in annotations.keys() and type(annotations.get(arg)) is str:
+                    print_fn_err("Do not document %s in both the docstring and annotations" % arg, parent, fn, errors)
+
+                params.append(arg)
+
+            # :type arg: type
+            match = type_re.match(line)
+            if match:
+                arg = match.group(1)
+                if arg in annotations.keys() and type(annotations.get(arg)) is type:
```

Comment:
Never do `type(...) is ...`, that will almost certainly not do what you want. Use `isinstance`.